### PR TITLE
lights out revamp (gold) 

### PR DESCRIPTION
**Pull request by ryanchou-dev**
image

---
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
tabs

**Comment by ryanchou-dev** [comment]
VSCODE IS THROWING

File: solutions/usaco-599.mdx [hunk 1]
```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
...
-const ll X = 97;
-const ll M = 999999929;
+const ll P = 31;
+const ll MOD = 1e9 + 7;
+const ll MX = 1e5 * 2;
 
-int n, m, H[505], p[505];
-vector<int> k;
-pii arr[205];
+struct coord {
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:50]: rename to `Coord`

File: solutions/usaco-599.mdx [hunk 2]
```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
...
-	p[0] = 1;
-	for(int i = 1; i < m; i++)
-	{
-		p[i] = int(p[i-1]*X%M);
-	}
-	H[0] = k[0];
-	for(int i = 1; i < m; i++)
-	{
-		H[i] = int((H[i-1]*X+k[i]) % M);
+signed main() {
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:54]:
```suggestion
int main() {
```

File: solutions/usaco-599.mdx [hunk 3]
```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
...
+	freopen("lightsout.in", "r", stdin);
+	freopen("lightsout.out", "w", stdout);
+
+	cin.tie(0)->sync_with_stdio(0);
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:58]: remove if runs w/o

File: solutions/usaco-599.mdx [hunk 4]
```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
...
-		{
-			cwdist[i][(j+1)%n] = cwdist[i][j] + abs((arr[j].f - arr[(j+1)%n].f) + (arr[j].s - arr[(j+1)%n].s));
-			if(j+1 != n) cwdist[(j+1)%n][i] = INT_MAX;
+	function<int(int i, int j)> calc_dist = [&](int i, int j) {
+		return abs(dt[i].x - dt[j].x) + abs(dt[i].y - dt[j].y);
+	};
+
+	vector<vector<vector<int>>> dst(n,
+																	vector<vector<int>>(n,
+																											vector<int>(2, INT32_MAX)));
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:74]: lol what

File: solutions/usaco-599.mdx [hunk 5]
```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
...
+
+	// i, j, 0 - clockwise dist from i -> j
+	// i, j, 1 - counter-clockwise dist from i -> j
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:77]: are these array indices or smth?

File: solutions/usaco-599.mdx [hunk 6]
```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
...
+		 * | ˅                     ˅  |
+		 * |--------------------------|
+		 *
+		 * |-------------------------|
+		 * |     angles of 270°      |
+		 * |              > ----------
+		 * |                |
+		 * |                |
+		 * |-----------------
+		 */
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:122]: i'd just put this in the editorial somewhere

File: solutions/usaco-599.mdx [hunk 7]
```diff
@@ -31,127 +31,172 @@ a one-base polynomial hash to check if the substring appears only once. If we
...
+
+		int strt = (i + 2) / 2;
+		if (fin >= n) {
+			fin = -1;
 		}
-		int b = dist[(sp+2)/2];
-		// cerr << a << " " << b << endl;
-		ans = max(ans, a-b);
+
+		int a = dst[strt][0][0];
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:188]: please don't remove vowels from the var names

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-599.mdx (hunk 5 above)
Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:77]: yeaaaaah

File: solutions/usaco-599.mdx (hunk 7 above)
Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:188]: ☃️

File: solutions/usaco-599.mdx (hunk 4 above)
Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:74]: ruh roh

File: solutions/usaco-599.mdx (hunk 6 above)
Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:122]: ill draw it up soon

File: solutions/usaco-599.mdx [hunk 8]
```diff
@@ -25,133 +25,169 @@ Bessie's location if the current path is not repeated somewhere else in the map.
...
-typedef pair<int, int> pii;
-#define f first
-#define s second
 using ll = long long;
 
-const ll X = 97;
-const ll M = 999999929;
+const ll P = 31;
+const ll MOD = 1e9 + 7;
+const ll MX = 1e5 * 2;
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:52]:
```suggestion
const int MX = 2e5;
```
or just use MAX

File: solutions/usaco-599.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/usaco-599.mdx:58]: not committed?

File: solutions/usaco-599.mdx [hunk 9]
```diff
@@ -25,133 +25,169 @@ Bessie's location if the current path is not repeated somewhere else in the map.
...
-		p[i] = int(p[i-1]*X%M);
-	}
-	H[0] = k[0];
-	for(int i = 1; i < m; i++)
-	{
-		H[i] = int((H[i-1]*X+k[i]) % M);
-	}
-}
+signed main() {
+  freopen("lightsout.in", "r", stdin);
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:59]: bru still tabs

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-599.mdx (hunk 2 above)
Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:58]: https://github.com/cpinitiative/usaco-guide/pull/3092/commits/5b69ddc95142ea333ef74de4abcd0b19e822accb

Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:58]: oops i think i undo'd it

File: solutions/usaco-599.mdx [hunk 10]
```diff
@@ -25,133 +25,169 @@ Bessie's location if the current path is not repeated somewhere else in the map.
...
+  vector<Coord> dt(n);
+  for (int i = 0; i < n; i++) {
+	cin >> dt[i].x >> dt[i].y;
+  }
+
+  function<int(int i, int j)> calc_dist = [&](int i, int j) {
+	return abs(dt[i].x - dt[j].x) + abs(dt[i].y - dt[j].y);
+  };
+
+  vector<vector<vector<int>>> dst(n,
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:74]: ??? is your ide dying or smth

Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:74]: fsdafasdfasdf

File: solutions/usaco-599.mdx [hunk 11]
```diff
@@ -25,133 +25,169 @@ Bessie's location if the current path is not repeated somewhere else in the map.
...
-void gen_dist()
-{
-	for(int i = 0; i < n; i++)
-	{
-		for(int j = i; j < n; j++)
-		{
-			cwdist[i][(j+1)%n] = cwdist[i][j] + abs((arr[j].f - arr[(j+1)%n].f) + (arr[j].s - arr[(j+1)%n].s));
-			if(j+1 != n) cwdist[(j+1)%n][i] = INT_MAX;
+	function<int(int i, int j)> calc_dist = [&](int i, int j) {
+	return abs(dt[i].x - dt[j].x) + abs(dt[i].y - dt[j].y);
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:71]:
```suggestion
		return abs(dt[i].x - dt[j].x) + abs(dt[i].y - dt[j].y);
```

File: solutions/usaco-599.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/usaco-599.mdx:81]: can you change it so that's more explicit

File: solutions/usaco-599.mdx [hunk 12]
```diff
@@ -25,133 +25,169 @@ Bessie's location if the current path is not repeated somewhere else in the map.
...
+	for (int i = 0; i < n; i++) {
+	for (int j = i; j < i + n; j++) {
+		if (i == j) {
+		dst[i][j][0] = 0;
+		} else {
+		int x = (j + n) % n;
+		int y = (j - 1 + n) % n;
+		dst[i][x][0] = dst[i][y][0] + calc_dist(x, y);
+		}
+	}
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:91]: sus

File: solutions/usaco-599.mdx [hunk 13]
```diff
@@ -25,133 +25,169 @@ Bessie's location if the current path is not repeated somewhere else in the map.
...
+	}
+	for (int i = 0; i < n; i++) {
+	for (int j = i; j > i - n; j--) {
+		if (i == j) {
+		dst[i][j][1] = 0;
+		} else {
+		int x = (j + n) % n;
+		int y = (j + 1 + n) % n;
+		dst[i][x][1] = dst[i][y][1] + calc_dist(x, y);
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:100]: even sussier

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-599.mdx (hunk 13 above)
Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:100]: aaaaaaaaaaaaaa

File: solutions/usaco-599.mdx [hunk 14]
```diff
@@ -25,133 +25,169 @@ Bessie's location if the current path is not repeated somewhere else in the map.
 If we were to frame this as a string-matching problem, we would want to find the
 shortest substring of the map such that it only appears once.
 
+To handle angles, we can note that there are only two possible interior angles, $90^\circ$ and $270^\circ$.
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-599.mdx:28]: do we not have $\degrees$?

Comment (Aadit-Ambadkar) [inline solutions/usaco-599.mdx:28]: do we not have $\degree$ (`$\degree$`)?

Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:28]: not sure. shouldn't really matter thougu

File: solutions/usaco-599.mdx [hunk 15]
```diff
@@ -25,133 +25,169 @@ Bessie's location if the current path is not repeated somewhere else in the map.
...
+const ll P = 31;
+const ll MOD = 1e9 + 7;
+const int MX = 2e5;
 
-int n, m, H[505], p[505];
-vector<int> k;
-pii arr[205];
+struct Coord {
+	int x, y;
+};
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:56]:
```suggestion
struct Coord { int x, y; };
```

File: solutions/usaco-599.mdx [hunk 16]
```diff
@@ -25,133 +25,169 @@ Bessie's location if the current path is not repeated somewhere else in the map.
...
+	};
+
+	vector<vector<vector<int>>> dst(n,
+		vector<vector<int>>(n,
+			vector<int>(2, INT32_MAX)
+		)
+	);
+
+	// index i, index j, 0 - clockwise dist from i -> j
+	// index i, index j, 1 - counter-clockwise dist from i -> j
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:81]:
```suggestion
	// dst[i][j][0] - clockwise dist from i -> j
```
same for the other
also please rename to `dist`

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-599.mdx (hunk 16 above)
Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:79]: ill fix the other part when i get home 👻

File: solutions/usaco-599.mdx [hunk 17]
```diff
@@ -25,133 +25,165 @@ Bessie's location if the current path is not repeated somewhere else in the map.
...
-typedef pair<int, int> pii;
-#define f first
-#define s second
 using ll = long long;
 
-const ll X = 97;
-const ll M = 999999929;
+const ll P = 31;
+const ll MOD = 1e9 + 7;
+const int MX = 2e5;
```
Comment (SansPapyrus683) [inline solutions/usaco-599.mdx:52]: don't see a 2e5 anywhere in the problem...can you clarify what this means?

Comment (ryanchou-dev) [reply solutions/usaco-599.mdx:53]: okok

**Review by envyaims — APPROVED** [review]
i like

**Pull request merged**